provisioning/device/src/Transports/Mqtt/ProvisioningTransportHandlerMqtt.cs

```diff
@@ -61,279 +64,315 @@ public class ProvisioningTransportHandlerMqtt : ProvisioningTransportHandler
+
+            if (Logging.IsEnabled)
+                Logging.Info(this, "Successfully sent the initial registration request, now polling until provisioning has finished.");
+
+            return await PollUntilProvisionigFinishes(mqttClient, registrationStatus.OperationId, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<RegistrationOperationStatus> PublishRegistrationRequest(IMqttClient mqttClient, ProvisioningTransportRegisterRequest provisioningRequest, CancellationToken cancellationToken)
+        {
+            DeviceRegistration registrationRequest = new DeviceRegistration()
+            {
+                Payload = new JRaw(provisioningRequest.Payload),
+            };
 
-                return operation.RegistrationState;
+            byte[] payload = new byte[0];
+
+            if (provisioningRequest.Payload != null)
+            {
+                string requestString = JsonConvert.SerializeObject(registrationRequest);
+                payload = Encoding.UTF8.GetBytes(requestString);
             }
-            catch (Exception ex) when (!(ex is ProvisioningTransportException))
+
+            int packetId = GetNextPacketId();
+            string registrationTopic = string.Format(CultureInfo.InvariantCulture, RegisterTopic, packetId);
+            var message = new MqttApplicationMessageBuilder()
+                .WithPayload(payload)
+                .WithTopic(registrationTopic)
+                .Build();
+
+            _startProvisioningRequestStatusSource = new TaskCompletionSource<RegistrationOperationStatus>();
+
+            try
             {
-                if (Logging.IsEnabled)
-                    Logging.Error(this, $"{nameof(ProvisioningTransportHandlerMqtt)} threw exception {ex}", nameof(RegisterAsync));
+                MqttClientPublishResult publishResult = await mqttClient.PublishAsync(message, cancellationToken).ConfigureAwait(false);
 
-                throw new ProvisioningTransportException($"MQTT transport exception", ex, true);
+                if (publishResult.ReasonCode != MqttClientPublishReasonCode.Success)
+                {
+                    throw new ProvisioningTransportException($"Failed to publish the MQTT packet for message with reason code {publishResult.ReasonCode}");
+                }
             }
-            finally
+            catch (Exception e) when (e is not ProvisioningTransportException)
             {
-                if (Logging.IsEnabled)
-                    Logging.Exit(this, $"{nameof(ProvisioningTransportHandlerMqtt)}.{nameof(RegisterAsync)}");
+                throw new ProvisioningTransportException("Failed to send the ?????", e);
             }
-        }
 
-        /// <inheritdoc/>
-        protected override void Dispose(bool disposing)
-        {
-            if (disposing)
+            //TODO cancellation token?
```

Comment:
AFAIK, the best you can do is Task.Delay(-1, cancellationToken), and await both tasks with Task.WhenAny.